src/hipscat_import/margin_cache/margin_cache.py

```diff
@@ -0,0 +1,249 @@
+                dec_column=args.catalog.catalog_info.dec_column,
+            )
+        )
+
+    for _ in tqdm(
+        as_completed(futures, with_results=True),
+        desc="Mapping  ",
+        total=len(futures),
+    ):
+        ...
+
+def _map_pixel_shards(
+    partition_file,
+    margin_pairs,
+    margin_threshold,
+    output_path,
+    margin_order,
+    ra_column,
+    dec_column
+):
+    """Creates margin cache shards from a source partition file."""
+    data = pd.read_parquet(partition_file)
+
+    data["margin_pixel"] = hp.ang2pix(
+        2**margin_order,
+        data[ra_column].values,
+        data[dec_column].values,
+        lonlat=True,
+        nest=True,
+    )
+
+    constrained_data = data.merge(margin_pairs, on="margin_pixel")
+
+    if len(constrained_data):
+        constrained_data.groupby(["partition_order", "partition_pixel"]).apply(
+            _to_pixel_shard,
+            margin_threshold=margin_threshold,
+            output_path=output_path,
+            margin_order=margin_order,
+            ra_column=ra_column,
+            dec_column=dec_column,
+        )
+
+def _to_pixel_shard(data, margin_threshold, output_path, margin_order, ra_column, dec_column):
+    """Do boundary checking for the cached partition and then output remaining data."""
+    order, pix = data["partition_order"].iloc[0], data["partition_pixel"].iloc[0]
+    source_order, source_pix = data["Norder"].iloc[0], data["Npix"].iloc[0]
+
+    scale = pixel_math.get_margin_scale(order, margin_threshold)
+    bounding_polygons = pixel_math.get_margin_bounds_and_wcs(order, pix, scale)
+    is_polar, _ = pixel_math.pixel_is_polar(order, pix)
+
+    if is_polar:
+        data = _margin_filter_polar(
+            data,
+            order,
+            pix,
+            margin_order,
+            margin_threshold,
+            ra_column,
+            dec_column,
+            bounding_polygons,
+        )
+    else:
+        data["margin_check"] = pixel_math.check_margin_bounds(
+            data[ra_column].values,
+            data[dec_column].values,
+            bounding_polygons
+        )
+
+    # pylint: disable-next=singleton-comparison
+    margin_data = data.loc[data["margin_check"] == True]
+
+    if len(margin_data):
+        # TODO: this should be a utility function in `hipscat`
+        # that properly handles the hive formatting
+        # generate a file name for our margin shard
+        partition_file = paths.pixel_catalog_file(output_path, order, pix)
+        partition_dir = f"{partition_file[:-8]}/"
+        shard_dir = paths.pixel_directory(
+            partition_dir, source_order, source_pix
+        )
+
+        file_io.make_directory(shard_dir, exist_ok=True)
+
+        shard_path = paths.pixel_catalog_file(
+            partition_dir, source_order, source_pix
+        )
+
+        final_df = margin_data.drop(columns=["partition_order", "partition_pixel", "margin_check"])
```

Comment:
Between the propagation of additional columns in the source data frame, and potential performance problems with exploding dataframes in this way, I'm wondering if it would be better to switch the logic to create a numpy mask, per margin pixel.

Reply:
`margin_check` is always added, good call on it not being removed. we can conditionally remove `is_trunc` as well

Reply:
also I thought about the numpy mask idea a little more and it _might_ be possible to do but would be kind of rough and I'm not sure it would actual provide much use value for our case. we can't really mask on margin pixel because the pairings are many-to-many, so we could mask on which `margin_pixel`s belong to _a_ `partition_pixel` but not what pixel they should go to. I guess we could keep a mask for each partition pixel but that kinda makes our memory problem worse 😭